aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -324,201 +315,120 @@ impl Aggregator {
-                            addition(current_delta, input, self.max_value)
-                                .expect("Addition is expected to succeed as the sum <= max_value"),
-                        )
-                    },
-                    DeltaValue::Negative(current_delta) => {
-                        // If speculative_start_value - current_delta + input is greater than max_value or less than -max_value, it's an overflow.
-                        // Otherwise, we update the delta with delta + input.
-                        if speculative_start_value.get_value() + input < current_delta {
-                            // speculative_start_value - current_delta + input < 0
-                            DeltaValue::Negative(subtraction(current_delta, input).expect(
-                                "Subtraction of smaller value from larger value must succeed",
-                            ))
-                        } else {
-                            // speculative_start_value - current_delta + input > 0
-                            if speculative_start_value.get_value() + input - current_delta
-                                > self.max_value
-                            {
-                                self.get_mut_history()
-                                    .unwrap()
-                                    .record_overflow(input - current_delta);
-                                return Err(abort_error(
-                                    format!("overflow occurred when adding {} to speculative_start_value: {:?}, current_delta: -{}", input, speculative_start_value, current_delta),
-                                    EADD_OVERFLOW,
-                                ));
-                            }
-                            if input > current_delta {
-                                DeltaValue::Positive(subtraction(input, current_delta).expect(
-                                    "Subtraction of smaller value from larger value must succeed",
-                                ))
-                            } else {
-                                DeltaValue::Negative(subtraction(current_delta, input).expect(
-                                    "Subtraction of smaller value from larger value must succeed",
-                                ))
-                            }
-                        }
-                    },
-                };
-                self.state = AggregatorState::Delta {
-                    speculative_start_value,
-                    delta: new_delta,
-                    history,
-                };
-                self.get_mut_history().unwrap().record_success(new_delta);
+                let cur_value = expect_ok(addition_deltavalue(speculative_start_value.get_value(), delta, self.max_value))?;
+
+                if addition(cur_value, input, self.max_value).is_err() {
+                    let overflow_value = expect_ok(ok_overflow(addition_deltavalue(input, delta, self.max_value)))?;
```

Comment:
Why `expect_ok` on top of `ok_underflow` here? Having an overflow here is not an invariant error.

Reply:
in this if, we are in the case that try_add exceeds limit (i.e. the is_err() check above)

and are now computing at what delta that happens, and so are computing  overflow_value = input + delta.

even if delta is negative, this cannot be < 0, and so we want to throw CODE_INVARIANT if that somehow ended up being < 0. 
if it is > limit, then we just don't want to record it, so we want overflow_value to be None in that case.

That's why we do ok_overflow - to transform overflow to None, and then expect_ok to raise CODE_INVARIANT.